[aptos-node] Enable consensus-only aptos-node for forge tests

ibalajiarun opened the pull request:
### Description

This PR enables running aptos-node in consensus-only mode by disabling execution and parts of mempool transaction validation features and by faking database calls. This is useful for evaluating the performance of the consensus stack individually in forge, and is not for *production use*. The change is feature-gated across multiple crates, such that without the feature enabled, the change is a no-op and should not affect **production binary** in any way.

Below is the description of changes per component. Each one of the change is gated behind a feature flag that should be enabled.
* Consensus: No change to how consensus communicates with the rest of the system
* Execution: Only the genesis block is executed on the VM while for other blocks an empty WriteSet is returned.
* Storage: Introduces `FakeAptosDB` that wraps AptosDB and also maintains some transaction and account related state values in-memory to support the transaction emitter and State Sync. `FakeAptosDB` overrides certain methods and serves them with in-memory data, and inherits the behavior for AptosDB for other calls. The genesis transactions are persisted to AptosDB.
* Mempool: The transaction validation mechanism is disabled and the transactions are added to the mempool directly. 
* State-sync: As is except that it is run in Transaction execution only mode with proof verification check disabled upon chunk execution.

Tests/CI Integration
- There is a new validator image that gets build with the feature if the CI `run-consensus-only-perf-test` label is set.
- There is a smoke test that runs as part of the GH workflow with the feature enabled
- There is a forge test that uses the new validator image and run if the CI label is set
- New unit tests have been added that tests when feature is enabled. Some existing tests that aren't necessary are ignored when feature is enabled.

### Test Plan
<!-- Please provide us with clear details for verifying that your changes work. -->

- Smoke test has been added with the feature enabled and integrated with CI
- Forge test runs with appropriate label
- New and existing tests

#5750

bchocho commented:
The approach seems reasonable, but let’s really put some effort in to clean it up.
* I’d hope lib.rs can end up being only a few lines of differences.
* Ideally, we can move the feature-gated parts to separate files where possible. (lib.rs is already huge as-is)

In order for this to not break, we probably need to add a smoke test that has the feature turned on.

Could we call the feature `consensus-only-perf-test` or something that is a little bit more descriptive?

Ideally, this doesn't add (much) burden when making changes to the actual prod code.
@grao1991 @JoshLind could you folks take a look from the storage and state sync side? Would it make future changes more difficult?

JoshLind commented on aptos-node/src/bootstrap.rs:
```diff
@@ -0,0 +1,56 @@
+use aptos_storage_interface::DbReaderWriter;
+use tokio::runtime::Runtime;
+
+#[cfg(not(feature = "consensus-only"))]
```
Can we not inline the feature gating (inside the method)? It seems like there's only a small difference between these two?

JoshLind commented on aptos-node/src/lib.rs:
```diff
@@ -382,6 +383,59 @@ fn fetch_chain_id(db: &DbReaderWriter) -> anyhow::Result<ChainId> {
+        .disable_lifo_slot()
+        .enable_all()
+        .build()
+        .map_err(|err| anyhow!("Failed to create data streaming service {}", err))?;
```
Nit: Update this comment 😄

JoshLind commented on mempool/src/shared_mempool/tasks.rs:
```diff
@@ -234,6 +232,34 @@ pub(crate) fn update_ack_counter(
 
+#[cfg(feature = "no-validation")]
+/// Submits a list of SignedTransaction to the local mempool
+/// and returns a vector containing AdmissionControlStatus.
```
nit: AdmissionControlStatus no longer exists 😄 Can we just delete this comment?

JoshLind commented on storage/aptosdb/src/fake_aptosdb.rs:
```diff
@@ -0,0 +1,555 @@
+                    latest_in_memory_state,
+                )?;
+
+                // let last_version = first_version + txns_to_commit.len() as u64 - 1;
```
nit: can we remove unused code?

ibalajiarun commented on storage/aptosdb/src/fake_aptosdb.rs:
```diff
@@ -0,0 +1,555 @@
+        let account_address = access_path.address;
+        let struct_tag = access_path.get_struct_tag();
+
+        if (account_address != aptos_test_root_address() && account_address != AccountAddress::ONE)
```
Sure, forge seems to checks the account balance for which it needs the coin resource, so it access the coin state value. We have that information the AptosDB because we wrote the whole writeset from actual genesis execution there, so we check and go to DB. 

Also, added code comments.

ibalajiarun commented on execution/executor/Cargo.toml:
```diff
@@ -52,3 +52,4 @@ vm-genesis = { workspace = true }
 default = []
 fuzzing = ["aptos-consensus-types/fuzzing", "aptos-crypto/fuzzing", "aptos-types/fuzzing", "storage-interface/fuzzing"]
 failpoints = ["fail/failpoints", "aptos-vm/failpoints"]
+fake-execute = []
```
Sure, I thought so too, but I was also thinking if we may need it in the future for something else, but what you say makes sense. We can keep it simple now and add in the future if needed.

ibalajiarun commented:
@bchocho @grao1991 @JoshLind  Thanks for taking a look. I have addressed you feedback and cleaned the code up.

It is also integrated with forge and there is a cluster perf test that runs in consensus-only mode if the appropriate label is set, which is set on this PR. :)

- I've also called the feature uniformly across all crates as `consensus-only-perf-test`
- There is a smoke test that runs in CI with this feature enabled.
- I tried following @JoshLind's idea of having the feature changes in a separate crate, but I ran into circular dependency problem with cargo, where the feature crate depends on mempool for example which in turn depends on the crate itself. If there can be a workaround or a different way to do this, please let me know, I will try.

JoshLind commented on execution/executor/src/chunk_executor.rs:
```diff
@@ -295,6 +285,68 @@ impl<V: VMExecutor> ChunkExecutorInner<V> {
+        Vec<Transaction>,
+    ),
+    anyhow::Error,
+> {
```
nit: maybe add a comment in here explicitly calling out that no validation is being done

JoshLind commented on execution/executor/src/tests/mod.rs:
```diff
@@ -145,15 +145,31 @@ fn test_executor_status() {
-        ],
-        output.compute_status()
-    );
+    #[cfg(not(feature = "consensus-only-perf-test"))]
```
nit: this feels kinda messy. Why not just ignore the entire test (like you've done above)?

JoshLind commented on execution/executor/src/tests/mod.rs:
```diff
@@ -213,7 +229,21 @@ fn test_executor_two_blocks_with_failed_txns() {
         .unwrap();
-    let ledger_info = gen_ledger_info(77, output2.root_hash(), block2_id, 1);
+
+    #[cfg(not(feature = "consensus-only-perf-test"))]
```
Same here. I'd just ignore the entire test. This kind of compilation flagging in tests is probably not worth it.

JoshLind commented on execution/executor/src/tests/mod.rs:
```diff
@@ -525,12 +556,19 @@ fn run_transactions_naive(transactions: Vec<Transaction>) -> HashValue {
     let db = &executor.db;
     let mut ledger_view: ExecutedTrees = db.reader.get_latest_executed_trees().unwrap();
 
+    #[cfg(not(feature = "consensus-only-perf-test"))]
```
Same here. I'd just ignore the entire test.

JoshLind commented on mempool/src/shared_mempool/tasks.rs:
```diff
@@ -350,6 +351,87 @@ where
+/// Submits a list of SignedTransaction to the local mempool
+/// and returns a vector containing SubmissionStatusBundle.
+#[cfg(feature = "consensus-only-perf-test")]
+pub(crate) fn process_incoming_transactions<V>(
```
nit: can we reduce the duplication between these two? e.g., define an internal helper that does most of the identical logic?

JoshLind commented on mempool/src/tests/shared_mempool_test.rs:
```diff
@@ -99,3 +100,53 @@ fn test_mempool_notify_committed_txns() {
+        kept_txn.clone(),
+    ];
+    // Add txns to mempool
+    {
```
nit: scoping feels odd to me.

ibalajiarun replied:
Copied from previous test as well. I think its for the lock, because its being held twice.

JoshLind commented on testsuite/smoke-test/src/consensus/consensus_only.rs:
```diff
@@ -0,0 +1,32 @@
+
+#[ignore]
+#[tokio::test]
+// Assumes that the
```
? Still a WIP? 😄

ibalajiarun replied:
Fixed it.

JoshLind commented on storage/aptosdb/src/fake_aptosdb.rs:
```diff
@@ -0,0 +1,894 @@
+use std::sync::Arc;
+
+/// Provides a "fake" AptosDB.
+pub struct FakeAptosDB {
```
It's hard to tell how much of this file is different to the regular DB? e.g., there appears to be a bunch of copy/pasted code, right? Is there any way to shrink this down, e.g., avoid the duplication?

ibalajiarun replied:
Ideally, we would want to "extend" AptosDB and just override the necessary methods, but I dont think it is possible in rust. I tried bunch of other things like implementing Deref, but it dint work because we are implementing multiple traits. I wanted to keep AptosDB untouched so this was an easy route. Let me know if there is anything else I should explore instead to make this better?

JoshLind replied:
Yeah, it's annoying that rust doesn't allow what you want to do... I'd check with @aptos-labs/storage. It might be better to just pull the common methods into a shared file.

msmouse replied:
It can be a hell for maintenance like this but it's cleanest from the normal path point of view. 
To make things more right, we might need a pretty large refactor which extracts out interfaces for 1. storing factual data (via RocksDB vs Memory) and 2. authenticating the state (alternative to SMT/JMT ). But I don't think it's viable in short term or worth it. cc @zjma in case you think differently.

Based on that I'm gonna approve.

msmouse replied:
It can be a hell for maintenance like this but it's cleanest from the normal path point of view.
To make things more right, we might need a pretty large refactor which extracts out interfaces for 1. storing factual data (via RocksDB vs Memory) and 2. authenticating the state (alternative to SMT/JMT ). But I don't think it's viable in short term or worth it. cc @zjma in case you think differently.

Based on that I'm gonna approve.

JoshLind reviewed (commented):
Thanks @ibalajiarun! 😃 The approach looks pretty reasonable to me. My only concerns are nits around readability and testing (e.g., duplicate code and whether we can just ignore most of the tests that diverge?). 

Given that most of this PR impacts the executor and storage code, I think they will ultimately need to sign-off. There are some non-trivial maintenance burdens from their side. 'cc @msmouse, @lightmark, @grao1991. Other than that, it LGTM 😄 .

ibalajiarun commented on execution/executor/src/tests/chunk_executor_tests.rs:
```diff
@@ -284,3 +286,50 @@ fn test_executor_execute_and_commit_chunk_local_result_mismatch() {
+    } = TestExecutor::new();
+
+    // commit 5 txns first.
+    {
```
This was carried over when copying from the previous test. There don't seem to be any Drop impl to warrant this, but I notice there are two ledger_info defitions, one inside the scope and one outside.

ibalajiarun commented on crates/transaction-emitter-lib/src/emitter/mod.rs:
```diff
@@ -275,7 +275,7 @@ impl EmitJobRequest {
                 // transactions, at least to the seconds granularity, so we reduce transactions_per_account
                 // if needed.
-                let transactions_per_account = min(20, tps);
+                let transactions_per_account = min(200, tps);
```
Not really, let me follow this up in a different one with some refactorings.

bchocho commented:
nit: the description says "consensus-only mode by disabling execution and state-sync" but state-sync is actually run as-is

bchocho commented:
FYI lint failures

bchocho commented on aptos-node/src/bootstrap.rs:
```diff
@@ -0,0 +1,57 @@
+    use aptos_backup_service::start_backup_service;
+
+    let (aptos_db, db_rw) = DbReaderWriter::wrap(
+        AptosDB::open(
```
Could this remove the repeated open?
https://gist.github.com/bchocho/7a8d5c17c718e7322cab95d6fa7d3189

ibalajiarun replied:
Thanks Brian, it works. I incorporated it.

bchocho commented on execution/executor/src/tests/mod.rs:
```diff
@@ -24,7 +24,6 @@ use aptos_types::{
     block_info::BlockInfo,
     chain_id::ChainId,
     ledger_info::{LedgerInfo, LedgerInfoWithSignatures},
-    proof::definition::LeafCount,
```
Any particular reason the imports were removed and fully-qualified names used?

ibalajiarun replied:
Fixed it. I was feature gating the test before, but ended up using cfg_attr with ignore because it was neater.

bchocho commented on mempool/src/shared_mempool/tasks.rs:
```diff
@@ -357,6 +358,88 @@ where
 }
 
+/// Submits a list of SignedTransaction to the local mempool
+/// and returns a vector containing SubmissionStatusBundle.
```
nit: [SubmissionStatusBundle] to be consistent

bchocho replied:
But also, should we add comments on how the perf test implementation differs from prod, on all of the `#[cfg(feature = "consensus-only-perf-test")]` functions?

ibalajiarun replied:
Sure, I have added as much info as possible. Let me know if something is missing.

bchocho commented on testsuite/forge-cli/src/main.rs:
```diff
@@ -1022,13 +1136,11 @@ fn chaos_test_suite(duration: Duration) -> ForgeConfig<'static> {
-                    1000
-                },
-            )
+            SuccessCriteria::new(if duration > Duration::from_secs(1200) {
```
There are several places where the format is just changed slightly. I wonder if this is a CLion vs VSCode thing? Ideally the already large PR isn't polluted by these

ibalajiarun replied:
I hadn't enabled +nightly flag for my rustfmt. It should be fixed now.

bchocho reviewed (commented):
Probably the most annoying thing is that CLion by default assumes all features are on, so the actual prod code gated by `#[cfg(not(feature = "consensus-only-perf-test"))]` is greyed out. A couple clicks can change the toggle, but I feel there will be some confusion in the future. I don't see how to "fix" this though.

I'll take a closer look at the mempool-side changes before approving. Let's get an approval from the storage side as well.

bchocho commented on testsuite/testcases/src/load_vs_perf_benchmark.rs:
```diff
@@ -34,7 +34,9 @@ impl LoadVsPerfBenchmark {
     ) -> Result<SingleRunStats> {
         let rng = SeedableRng::from_rng(ctx.core().rng())?;
-        let emit_job_request = ctx.emit_job.clone().mode(EmitJobMode::ConstTps { tps });
+        let emit_job_request = ctx.emit_job.clone().mode(EmitJobMode::MaxLoad {
```
Do we want these changes as part of this PR?

ibalajiarun replied:
Undid it here. Let me do a follow up PR to get the best out of the consensus-only forge tests.

bchocho commented on aptos-node/src/storage.rs:
```diff
@@ -13,12 +12,39 @@ use aptos_config::{
+}
+
+#[cfg(feature = "consensus-only-perf-test")]
+pub(crate) fn bootstrap_db(
```
This one is missing a comment. Since FakeAptosDB is already well-documented, maybe just something simple like "Use an in-memory DB (FakeAptosDB) and avoid writing db backups."?

bchocho commented on testsuite/forge-cli/src/main.rs:
```diff
@@ -479,11 +479,105 @@ fn single_test_suite(test_name: &str) -> Result<ForgeConfig<'static>> {
+        }))
+        // TODO(rustielin): tune these success critiera after we have a better idea of the test behavior
+        .with_success_criteria(
+            SuccessCriteria::new(3000)
```
Can we bump up the success criteria?

ibalajiarun replied:
Bumping to 10k now but let me land this and follow up with another PR that tunes the forge runs.

bchocho reviewed (approved):
A few comments, but this is approved from my side (the global approach + Mempool)

zekun000 commented:
one thing I'm paranoid about is if we accidentally compile with the feature enabled implicitly, we had this problem before when fail_points was enabled implicitly with feature unifications for production binary.

maybe we can add a config value or pass in argument at main to ensure that it's aligned with the compile flags, like
```
  assert!(args.consensus_only == cfg!(feature="consensus-only-perf-test", "Misconfigured");
```
